File: Tree/BST.java
/* 
   This program demonstrates the implementation of a Binary Search Tree (BST) in C++.
   
   The program creates a BST and performs various operations on it, including:
   - Insertion: Adds new nodes to the tree.
   - Deletion: Removes nodes from the tree.
   - Searching: Finds specific nodes in the tree.
   - Traversal: Visits all nodes in the tree in specific orders (in-order, pre-order, post-order, level-order(BFS)) and DFS.
   - Height/Depth: Calculates the maximum number of edges from the root to a leaf.
   - Size: Counts the total number of nodes in the tree.
   - Minimum/Maximum: Finds the minimum or maximum value in the tree.

   The main function demonstrates the usage of these operations.

   Note: This is a basic implementation assuming unique values in the tree. Additional functionality 
   and error handling can be added as per specific requirements.
*/


import java.util.LinkedList;
import java.util.Queue;

class Node {
    int data;
    Node left;
    Node right;

    public Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BST {
    private Node root;

    public BST() {
        root = null;
    }

    public void insert(int data) {
        root = insertRecursive(root, data);
    }

    private Node insertRecursive(Node root, int data) {
        if (root == null) {
            root = new Node(data);
            return root;
        }

        if (data < root.data) {
            root.left = insertRecursive(root.left, data);
        } else if (data > root.data) {
            root.right = insertRecursive(root.right, data);
        }

        return root;
    }

    public void delete(int data) {
        root = deleteRecursive(root, data);
    }

    private Node deleteRecursive(Node root, int data) {
        if (root == null) {
            return root;
        }

        if (data < root.data) {
            root.left = deleteRecursive(root.left, data);
        } else if (data > root.data) {
            root.right = deleteRecursive(root.right, data);
        } else {
            if (root.left == null) {
                return root.right;
            } else if (root.right == null) {
                return root.left;
            }

            root.data = minValue(root.right);
            root.right = deleteRecursive(root.right, root.data);
        }

        return root;
    }

    private int minValue(Node root) {
        int minValue = root.data;
        while (root.left != null) {
            minValue = root.left.data;
            root = root.left;
        }
        return minValue;
    }

    public boolean search(int data) {
        return searchRecursive(root, data);
    }

    private boolean searchRecursive(Node root, int data) {
        if (root == null) {
            return false;
        }

        if (data == root.data) {
            return true;
        } else if (data < root.data) {
            return searchRecursive(root.left, data);
        } else {
            return searchRecursive(root.right, data);
        }
    }

    public void inOrderTraversal() {
        inOrderTraversalRecursive(root);
        System.out.println();
    }

    private void inOrderTraversalRecursive(Node root) {
        if (root != null) {
            inOrderTraversalRecursive(root.left);
            System.out.print(root.data + " ");
            inOrderTraversalRecursive(root.right);
        }
    }

    public void preOrderTraversal() {
        preOrderTraversalRecursive(root);
        System.out.println();
    }

    private void preOrderTraversalRecursive(Node root) {
        if (root != null) {
            System.out.print(root.data + " ");
            preOrderTraversalRecursive(root.left);
            preOrderTraversalRecursive(root.right);
        }
    }

    public void postOrderTraversal() {
        postOrderTraversalRecursive(root);
        System.out.println();
    }

    private void postOrderTraversalRecursive(Node root) {
        if (root != null) {
            postOrderTraversalRecursive(root.left);
            postOrderTraversalRecursive(root.right);
            System.out.print(root.data + " ");
        }
    }

    public void levelOrderTraversal() {
        if (root == null) {
            return;
        }

        Queue<Node> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            Node current = queue.poll();
            System.out.print(current.data + " ");

            if (current.left != null) {
                queue.add(current.left);
            }

            if (current.right != null) {
                queue.add(current.right);
            }
        }

        System.out.println();
    }

    public int getHeight() {
        return getHeightRecursive(root);
    }

    private int getHeightRecursive(Node root) {
        if (root == null) {
            return 0;
        }

        int leftHeight = getHeightRecursive(root.left);
        int rightHeight = getHeightRecursive(root.right);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    public int getSize() {
        return getSizeRecursive(root);
    }

    private int getSizeRecursive(Node root) {
        if (root == null) {
            return 0;
        }

        int leftSize = getSizeRecursive(root.left);
        int rightSize = getSizeRecursive(root.right);

        return leftSize + rightSize + 1;
    }

    public int getMinimum() {
        if (root == null) {
            throw new IllegalStateException("Tree is empty");
        }

        Node current = root;
        while (current.left != null) {
            current = current.left;
        }

        return current.data;
    }

    public int getMaximum() {
        if (root == null) {
            throw new IllegalStateException("Tree is empty");
        }

        Node current = root;
        while (current.right != null) {
            current = current.right;
        }

        return current.data;
    }
// }

// public class Main {
    public static void main(String[] args) {
        BST bst = new BST();
        bst.insert(50);
        bst.insert(30);
        bst.insert(70);
        bst.insert(20);
        bst.insert(40);
        bst.insert(60);
        bst.insert(80);

        System.out.println("In-order traversal:");
        bst.inOrderTraversal();

        System.out.println("Pre-order traversal:");
        bst.preOrderTraversal();

        System.out.println("Post-order traversal:");
        bst.postOrderTraversal();

        System.out.println("Level-order traversal:");
        bst.levelOrderTraversal();

        System.out.println("Height of the tree: " + bst.getHeight());

        System.out.println("Size of the tree: " + bst.getSize());

        System.out.println("Minimum value in the tree: " + bst.getMinimum());

        System.out.println("Maximum value in the tree: " + bst.getMaximum());

        int searchData = 60;
        if (bst.search(searchData)) {
            System.out.println(searchData + " is found in the tree.");
        } else {
            System.out.println(searchData + " is not found in the tree.");
        }

        bst.delete(30);

        System.out.println("In-order traversal after deleting 30:");
        bst.inOrderTraversal();
    }
}

File: Tree/BST.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    def insert(self, data):
        self.root = self._insert_recursive(self.root, data)

    def _insert_recursive(self, root, data):
        if root is None:
            return Node(data)

        if data < root.data:
            root.left = self._insert_recursive(root.left, data)
        elif data > root.data:
            root.right = self._insert_recursive(root.right, data)

        return root

    def delete(self, data):
        self.root = self._delete_recursive(self.root, data)

    def _delete_recursive(self, root, data):
        if root is None:
            return root

        if data < root.data:
            root.left = self._delete_recursive(root.left, data)
        elif data > root.data:
            root.right = self._delete_recursive(root.right, data)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left

            min_value = self._min_value(root.right)
            root.data = min_value
            root.right = self._delete_recursive(root.right, min_value)

        return root

    def _min_value(self, root):
        min_val = root.data
        while root.left is not None:
            min_val = root.left.data
            root = root.left
        return min_val

    def search(self, data):
        return self._search_recursive(self.root, data)

    def _search_recursive(self, root, data):
        if root is None:
            return False

        if data == root.data:
            return True
        elif data < root.data:
            return self._search_recursive(root.left, data)
        else:
            return self._search_recursive(root.right, data)

    def in_order_traversal(self):
        self._in_order_traversal_recursive(self.root)
        print()

    def _in_order_traversal_recursive(self, root):
        if root is not None:
            self._in_order_traversal_recursive(root.left)
            print(root.data, end=' ')
            self._in_order_traversal_recursive(root.right)

    def pre_order_traversal(self):
        self._pre_order_traversal_recursive(self.root)
        print()

    def _pre_order_traversal_recursive(self, root):
        if root is not None:
            print(root.data, end=' ')
            self._pre_order_traversal_recursive(root.left)
            self._pre_order_traversal_recursive(root.right)

    def post_order_traversal(self):
        self._post_order_traversal_recursive(self.root)
        print()

    def _post_order_traversal_recursive(self, root):
        if root is not None:
            self._post_order_traversal_recursive(root.left)
            self._post_order_traversal_recursive(root.right)
            print(root.data, end=' ')

    def level_order_traversal(self):
        if self.root is None:
            return

        queue = []
        queue.append(self.root)

        while queue:
            current = queue.pop(0)
            print(current.data, end=' ')

            if current.left is not None:
                queue.append(current.left)
            if current.right is not None:
                queue.append(current.right)

        print()

    def get_height(self):
        return self._get_height_recursive(self.root)

    def _get_height_recursive(self, root):
        if root is None:
            return 0

        left_height = self._get_height_recursive(root.left)
        right_height = self._get_height_recursive(root.right)

        return max(left_height, right_height) + 1

    def get_size(self):
        return self._get_size_recursive(self.root)

    def _get_size_recursive(self, root):
        if root is None:
            return 0

        left_size = self._get_size_recursive(root.left)
        right_size = self._get_size_recursive(root.right)

        return left_size + right_size + 1

    def get_minimum(self):
        if self.root is None:
            raise Exception("Tree is empty")

        current = self.root
        while current.left is not None:
            current = current.left

        return current.data

    def get_maximum(self):
        if self.root is None:
            raise Exception("Tree is empty")

        current = self.root
        while current.right is not None:
            current = current.right

        return current.data


bst = BST()
bst.insert(50)
bst.insert(30)
bst.insert(70)
bst.insert(20)
bst.insert(40)
bst.insert(60)
bst.insert(80)

print("In-order traversal:")
bst.in_order_traversal()

print("Pre-order traversal:")
bst.pre_order_traversal()

print("Post-order traversal:")
bst.post_order_traversal()

print("Level-order traversal:")
bst.level_order_traversal()

print("Height of the tree:", bst.get_height())

print("Size of the tree:", bst.get_size())

print("Minimum value in the tree:", bst.get_minimum())

print("Maximum value in the tree:", bst.get_maximum())

search_data = 60
if bst.search(search_data):
    print(search_data, "is found in the tree.")
else:
    print(search_data, "is not found in the tree.")

bst.delete(30)

print("In-order traversal after deleting 30:")
bst.in_order_traversal()
